probatus/utils/interface.py

```diff
@@ -0,0 +1,52 @@
+# Copyright (c) 2020 ING Bank N.V.
+#
+# Permission is hereby granted, free of charge, to any person obtaining a copy of
+# this software and associated documentation files (the "Software"), to deal in
+# the Software without restriction, including without limitation the rights to
+# use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of
+# the Software, and to permit persons to whom the Software is furnished to do so,
+# subject to the following conditions:
+#
+# The above copyright notice and this permission notice shall be included in all
+# copies or substantial portions of the Software.
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS
+# FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
+# COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
+# IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+# CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+
+
+from abc import ABC, abstractmethod
+from probatus.utils import NotFittedError
+
+
+class BaseFitComputeClass(ABC):
+    fitted = False
+
+    def _check_if_fitted(self):
+        """
+        Checks if object has been fitted. If not, NotFittedError is raised.
+        """
+        if self.fitted is False:
+            raise(NotFittedError('The object has not been fitted. Please run fit() method first'))
+
+    @abstractmethod
+    def fit(self, *args, **kwargs):
+        pass
+
+    @abstractmethod
+    def compute(self, *args, **kwargs):
+        pass
+
+    @abstractmethod
+    def fit_compute(self, *args, **kwargs):
```

Comment:
any reason you choose to make this an abstractmethod, forcing an implementation in each and every class that inherits from this baseclass? In most cases, the implementation is `return self.fit(X, y).compute()` and we could implement that in the baseclass.

One reason I would see for being explicit about `fit_compute` is the API documentation, because you want to show that a class also has a `fit_compute()` method. the mkdocstrings package has an issue for this (https://github.com/pawamoy/mkdocstrings/issues/157) and with some research it might be possible to specify which methods from the base class to include.

Reply:
There are two reasons why i want to overwrite it every time:
- Sometimes the compute methods takes an argument, and in this case we always have to overwrite it
- Overwriting it allows to be specific about the parameters of the function e.g. in ResemblanceModel it is X1 and X2, in ShapRFECV X, y. Moreover, we can nicely describe the inputs and the outputs in the docstring.

Having a non-abstract fit_compute in the interface gives a minimal gain, because in most cases you anyway want to describe the docstrings and show which parameters the user can input.

Reply:
Clear!